### PR TITLE
F5 framework selector

### PR DESCRIPTION
**Pull request by BillHiebert**
PR for post dev15 RTW.
Implemented the design that was available in Dev14:
1.	Added a new menu item to the debug dropdown which shows the currently active framework being used for debugging. 
a.	The selected framework appears in the Framework menu name so that you don’t have to expand the menu to see the active one.
b.	The frameworks appear in the menu in the order they are defined in the <TargetFrameworks> property
c.	The current active framework is remembered in the .user file so it persists across VS sessions. It defaults to the first one if there is no existing setting or the existing setting doesn’t match any the frameworks being targeted.
d.	The framework names are displayed as they exist in the project file. I’d like to display something better but I didn’t have any code to convert the short name to a friendly name.  We could possible resolve to the TargetFrameworkMoniker property and display that - but that requires more evaluation. Certainly something to address later.
2.	Added an IActiveDebugFrameworkServices public interface which exposes access to the ConfiguredProject for the active debug framework. 
3.	Updated the Debugger providers and (token replacer) to resolve properties against the active debug frameworks ConfiguredProject. 
4. Unit test coverage for the new components
5. Fixed a problem with ProjectExportProvider. It claims it can be called on a background thread but doing so deadlocked as it was using QueryService to get to ProjectServices. Fixed it to just import the ProjectServices MEF component.

@davkean @srivatsn @abpiskunov

### CODE REVIEW

**Comment by basoundr** [comment]
@BillHiebert You can probably create an issue to address 1-d when commit this change

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/DebugTokenReplacer.cs [hunk 1]
```diff
@@ -18,14 +18,17 @@ internal class DebugTokenReplacer : IDebugTokenReplacer
...
+                                  IActiveDebugFrameworkServices activeDebugFramework)
         {
             UnconfiguredServices = unconnfiguredServices;
             EnvironmentHelper = environmentHelper;
+            ActiveDebugFramework = activeDebugFramework;
         }
 
         private IUnconfiguredProjectCommonServices UnconfiguredServices { get; }
         private IEnvironmentHelper EnvironmentHelper { get; }
+        private IActiveDebugFrameworkServices ActiveDebugFramework { get; }
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/DebugTokenReplacer.cs]: Rename to `ActiveDebugFrameworkService`?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/DebugTokenReplacer.cs [hunk 2]
```diff
@@ -118,9 +121,11 @@ public DebugTokenReplacer(IUnconfiguredProjectCommonServices unconnfiguredServic
         /// This is here to support unit tests which can derive from this class and return their own 
         /// instance of IProjectReadAccess
         /// </summary>
-        protected virtual IProjectReadAccess AccessProject()
+        protected virtual async Task<IProjectReadAccess> AccessProject()
         {
-            return new ProjectReadAccessor(UnconfiguredServices.ProjectLockService, UnconfiguredServices.ActiveConfiguredProject);
+            var debuggingConfig = await ActiveDebugFramework.GetConfiguredProjectForActiveFrameworkAsync().ConfigureAwait(true);
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/DebugTokenReplacer.cs]: There is also a `ProjectConfig`. To make it unambiguous, rename this to `debuggingActiveConfiguredProject` or `debuggingConfiguredProject`?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/ActiveDebugFrameworkServices.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+            // We can't just select the first one. If activeFramework is not set we must pick the first one as defined by the 
+            // targetFrameworks property. So we need the order as returned by GetProjectFrameworks()
+            var frameworks = await GetProjectFrameworksAsync().ConfigureAwait(false);
+            if(frameworks != null &&  frameworks.Count > 0)
+            {
+                if(configProjects.TryGetValue(frameworks[0], out ConfiguredProject configuredProject))
+                {
+                    return configuredProject;
+                }
+
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/ActiveDebugFrameworkServices.cs]: Remove line break

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/ActiveDebugFrameworkServices.cs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            }
+
+            // All that is left is to return the first one.
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/ActiveDebugFrameworkServices.cs]: When do we get to this state? It seems like we would get here if TargetFramework or TargetFrameworks is not mentioned. Which would mean configProjects will be empty we will throw an exception at `configProjects.First()`

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/Debug/ActiveDebugFrameworkServicesTests.cs [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+                PropertyName = ProjectDebugger.ActiveDebugFrameworkProperty,
+                Value = "FrameworkOne"
+            };
+
+            var projectProperties = ProjectPropertiesFactory.Create(project, data);
+            
+            var commonServices = IUnconfiguredProjectCommonServicesFactory.Create(projectProperties: projectProperties);
+
+            var debugFrameworkSvcs = new ActiveDebugFrameworkServices(null, commonServices);
+            await debugFrameworkSvcs.SetActiveDebuggingFrameworkPropertyAsync("netcoreapp1.0");
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/Debug/ActiveDebugFrameworkServicesTests.cs]: test missing assert

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/Debug/ActiveDebugFrameworkServicesTests.cs [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+            
+            var commonServices = IUnconfiguredProjectCommonServicesFactory.Create(projectProperties: projectProperties);
+
+            var debugFrameworkSvcs = new ActiveDebugFrameworkServices(projectConfgProvider.Object, commonServices);
+            var activeConfiguredProject = await debugFrameworkSvcs.GetConfiguredProjectForActiveFrameworkAsync();
+            Assert.Equal(selectedConfigFramework,  activeConfiguredProject.ProjectConfiguration.Dimensions.GetValueOrDefault("TargetFramework"));
+
+        }
+
+        //[Fact]
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/Debug/ActiveDebugFrameworkServicesTests.cs]: forgot to delete this?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/ActiveDebugFrameworkServices.cs (hunk 4 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/ActiveDebugFrameworkServices.cs]: If there aren't any configurations I am pretty sure the project won't even load so throwing on .First() is probably the right thing to do here.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/Debug/ActiveDebugFrameworkServicesTests.cs (hunk 6 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/Debug/ActiveDebugFrameworkServicesTests.cs]: Yes.

**Comment by abpiskunov** [comment]
For this one 'The framework names are displayed as they exist in the project file. I’d like to display something better but I didn’t have any code to convert the short name to a friendly name. We could possible resolve to the TargetFrameworkMoniker property and display that - but that requires more evaluation. Certainly something to address later.'

I also need it in my pending changes and i will add a global component that would resolve actual TFMs using Nuget COM API. Just FYI , to avoid dupes...

**Pull request closed without merging**